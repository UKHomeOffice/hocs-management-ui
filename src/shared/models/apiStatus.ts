export default interface ApiStatus {
    display: string;
    timeoutPeriod: number;
    type: string;
}
