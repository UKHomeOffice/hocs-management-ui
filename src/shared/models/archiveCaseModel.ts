export default interface ArchiveCaseModel {
    caseReference: string;
    deleted: boolean;
}
