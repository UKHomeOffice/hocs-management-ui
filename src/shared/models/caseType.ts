export default interface CaseType {
    displayName: string;
    label: string;
    shortCode: string;
    type: string;
    value: string;
}
