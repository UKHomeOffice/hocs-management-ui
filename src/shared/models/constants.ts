export const ADD_CHILD_TOPIC_ERROR_DESCRIPTION = 'Something went wrong while adding the child topic. Please try again.';
export const ADD_STANDARD_LINE_ERROR_DESCRIPTION = 'Something went wrong while adding the standard line. Please try again.';
export const ADD_TEMPLATE_ERROR_DESCRIPTION = 'Something went wrong while adding the template. Please try again.';
export const ADD_UNIT_ERROR_DESCRIPTION = 'Something went wrong while adding the unit. Please try again.';
export const ADD_USER_ERROR_DESCRIPTION = 'Something went wrong while adding the following users. Please try again.';
export const ADD_USER_ERROR_TITLE = 'There was an error adding the users';
export const DUPLICATE_CHILD_TOPIC_DESCRIPTION = 'A child topic with that name already exists';
export const DUPLICATE_UNIT_DESCRIPTION = 'A unit with those details already exists';
export const EMPTY_SUBMIT_ERROR_DESCRIPTION = 'Please select some users before submitting.';
export const EMPTY_SUBMIT_ERROR_TITLE = 'No users selected';
export const EMPTY_SUBMIT_TEAM_ERROR_DESCRIPTION = 'Please select a team before submitting.';
export const EMPTY_SUBMIT_TEAM_ERROR_TITLE = 'No team selected';
export const EMPTY_SUBMIT_DRAFT_QA_TEAM_ERROR_DESCRIPTION = 'Select a team for Initial Draft and QA response stages.';
export const EMPTY_SUBMIT_PRIVATE_MINISTER_TEAM_ERROR_DESCRIPTION = 'Select a team for Private Office and Minister sign off stages.';
export const EMPTY_SUBMIT_TOPIC_ERROR_DESCRIPTION = 'Select a topic before submitting.';
export const EMPTY_SUBMIT_TOPIC_ERROR_TITLE = 'No topic selected';
export const GENERAL_ERROR_TITLE = 'Something went wrong';
export const LOAD_PARENT_TOPICS_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_TOPICS_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_TOPIC_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_CASE_TYPES_DESCRIPTION = 'There was an error retrieving the case types.  Please try refreshing the page.';
export const LOAD_TEAM_ERROR_DESCRIPTION = 'There was an error retrieving the team.  Please try refreshing the page.';
export const LOAD_TEAM_MEMBERS_ERROR_DESCRIPTION = 'There was an error retrieving the teams members.  Please try refreshing the page.';
export const LOAD_TEAMS_ERROR_DESCRIPTION = 'There was an error retrieving the teams.  Please try refreshing the page.';
export const LOAD_USERS_ERROR_DESCRIPTION = 'There was an error retrieving the list of users.  Please try refreshing the page.';
export const REMOVE_FROM_TEAM_ALLOCATED_ERROR_DESCRIPTION = 'The user cannot be removed from the team as they have cases assigned';
export const REMOVE_FROM_TEAM_ERROR_DESCRIPTION = 'There was an error removing the user from the team';
export const VALIDATION_ERROR_TITLE = "Thereâ€™s a problem";
