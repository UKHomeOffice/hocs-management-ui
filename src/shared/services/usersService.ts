import apiClient from './apiClient';
import Item from '../models/item';

const usersApi = apiClient.createClient();

export class AddUserError extends Error {
    userToAdd: Item;
    constructor(message: string, userToAdd: Item) {
        super(message);
        this.userToAdd = userToAdd;
    }
}

export const getUsers = () => usersApi.get('/api/users');

export const addUserToTeam = (user: Item, teamId: string) => {
    return new Promise((resolve, reject) => usersApi
        .post(`/api/users/${user}/team/${teamId}`)
        .then(value => resolve(value.data))
        .catch(reason => reject(new AddUserError(reason, user))));
};

export const deleteUserFromTeam = (userId: string, teamId: string) => usersApi.delete(`/api/users/${userId}/team/${teamId}`);
