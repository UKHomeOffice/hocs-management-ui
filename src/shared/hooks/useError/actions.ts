import { FormError } from '../../models/formError';

export type AddFormError = {
    payload: FormError;
    type: 'AddFormError';
};

export type ClearError = {
    type: 'ClearError'
};

export type SetTitle = {
    payload: string;
    type: 'SetTitle';
};

export type SetDescription = {
    payload: string;
    type: 'SetDescription';
};

export type Action =
    AddFormError |
    ClearError |
    SetDescription |
    SetTitle;
