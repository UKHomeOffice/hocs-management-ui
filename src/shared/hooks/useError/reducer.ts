import { Action } from './actions';
import PageError from '../../models/pageError';

export const reducer = (state: PageError, action: Action): PageError => {
    switch (action.type) {
        case 'AddFormError':
            return {
                error: { ...state.error, formErrors: [...state.error ? state.error.formErrors || [] : [], action.payload] }
            };
        case 'ClearError':
            return { error: undefined };
        case 'SetDescription':
            return { error: { ...state.error, description: action.payload } };
        case 'SetTitle':
            return { error: { ...state.error, title: action.payload } };
    }
    return state;
};
