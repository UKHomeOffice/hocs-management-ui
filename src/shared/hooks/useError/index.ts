import React, { Reducer } from 'react';
import PageError from '../../models/pageError';
import { Action } from './actions';
import { reducer } from './reducer';
import { FormError } from '../../models/formError';
import ErrorMessage from 'shared/models/errorMessage';

const useError = (formErrorDescription?: string, formErrorTitle?: string)
    : [
        PageError,
        (formError: FormError) => void,
        () => void,
        (error: ErrorMessage) => void
    ] => {
    const [pageError, dispatch] = React.useReducer<Reducer<PageError, Action>>(reducer, {});

    const addFormError = (formError: FormError) => {
        (formErrorDescription || formErrorTitle) && setMessage({ title: formErrorTitle || '', description: formErrorDescription || '' });
        dispatch({ type: 'AddFormError', payload: formError });
    };
    const clearErrors = () => dispatch({ type: 'ClearError' });
    const setMessage = (errorMessage: ErrorMessage) => {
        dispatch({ type: 'SetDescription', payload: errorMessage.description });
        dispatch({ type: 'SetTitle', payload: errorMessage.title });
    };
    return [pageError, addFormError, clearErrors, setMessage];
};

export default useError;
