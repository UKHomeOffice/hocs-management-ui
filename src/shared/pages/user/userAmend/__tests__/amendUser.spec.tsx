
import React from 'react';
import { match, MemoryRouter } from 'react-router-dom';
import { createBrowserHistory, History, Location } from 'history';
import { act, render, RenderResult, waitFor, fireEvent } from '@testing-library/react';
import { GENERAL_ERROR_TITLE, LOAD_USER_ERROR_DESCRIPTION, AMEND_USER_ERROR_DESCRIPTION, AMEND_USER_ERROR_TITLE, AMEND_USER_BAD_REQUEST_TITLE } from '../../../../models/constants';
import AmendUser from '../amendUser';
import * as useError from '../../../../hooks/useError';
import State from '../state';
import * as UsersService from '../../../../services/usersService';

let match: match<any>;
let history: History<any>;
let location: Location;
let mockState: State;
let wrapper: RenderResult;

const useReducerSpy = jest.spyOn(React, 'useReducer');
const reducerDispatch = jest.fn();
const useErrorSpy = jest.spyOn(useError, 'default');
const addFormErrorSpy = jest.fn();
const clearErrorsSpy = jest.fn();
const setMessageSpy = jest.fn();
const userId = 'xx-xx-xx-xx';

const renderComponent = () => render(
    <MemoryRouter>
        <AmendUser history={history} location={location} match={match}/>
    </MemoryRouter>
);
const getUserSpy = jest.spyOn(UsersService, 'getUser');
const updateUserSpy = jest.spyOn(UsersService, 'updateUser');
beforeEach(() => {
    history = createBrowserHistory();
    match = {
        isExact: true,
        params: { userUUID: userId },
        path: '',
        url: ''
    };

    location = {
        hash: '',
        key: '',
        pathname: '',
        search: '',
        state: {}
    };
    mockState = {
        username: '',
        email: '',
        firstName: '',
        lastName: '',
        enabled: 'true'
    };

    getUserSpy.mockImplementation(() => Promise.resolve({ id: 'id', username: 'un', email: 'email', firstName: 'testSimpleName', lastName: 'testTitle', uuid: 'testUUID', enabled: true }));
    updateUserSpy.mockImplementation(() => Promise.resolve());
    useReducerSpy.mockImplementation(() => [mockState, reducerDispatch]);
    useErrorSpy.mockImplementation(() => [{}, addFormErrorSpy, clearErrorsSpy, setMessageSpy]);
    history.push = jest.fn();
    reducerDispatch.mockReset();
    addFormErrorSpy.mockReset();
    clearErrorsSpy.mockReset();
    setMessageSpy.mockReset();
    act(() => {
        wrapper = renderComponent();
    });
});

describe('when the AmendUser component is mounted', () => {
    it('should render with default props', async () => {
        expect.assertions(7);
        wrapper = renderComponent();
        await waitFor(() => {
            expect(getUserSpy).toHaveBeenCalled();
            expect(reducerDispatch).toHaveBeenNthCalledWith(1, { name: 'username', value: 'un' });
            expect(reducerDispatch).toHaveBeenNthCalledWith(2, { name: 'email', value: 'email' });
            expect(reducerDispatch).toHaveBeenNthCalledWith(3, { name: 'firstName', value: 'testSimpleName' });
            expect(reducerDispatch).toHaveBeenNthCalledWith(4, { name: 'lastName', value: 'testTitle' });
            expect(reducerDispatch).toHaveBeenNthCalledWith(5, { name: 'enabled', value: 'true' });
            expect(wrapper.container).toMatchSnapshot();
        });
    });

    it('should display an error if the call to retrieve item details fails', async () => {
        expect.assertions(2);
        getUserSpy.mockImplementation(() => Promise.reject('error'));
        wrapper = renderComponent();
        await waitFor(() => {
            expect(setMessageSpy).toBeCalledWith({ title: GENERAL_ERROR_TITLE, description: LOAD_USER_ERROR_DESCRIPTION });
        });

    });
});

describe('when the new first name is entered', () => {
    it('should be persisted in the page state', async () => {
        wrapper = renderComponent();
        getUserSpy.mockReturnValueOnce(Promise.resolve(
            { id: 'id', username: 'un', email: 'email', firstName: 'testSimpleName', lastName: 'testTitle', uuid: 'testUUID', enabled: true }
        ));
        const nameElement = await waitFor(async () => {
            return await wrapper.findByLabelText('First Name');
        });
        const newFirstName = 'Xyz';
        fireEvent.change(nameElement, { target: { name: 'firstName', value: newFirstName } });
        await waitFor(() => {
            expect(reducerDispatch).toHaveBeenCalledWith({ name: 'firstName', value: newFirstName });
        });
    });
});

describe('when the submit button is clicked', () => {
    describe('and the data is filled in', () => {

        beforeEach(async () => {
            mockState.firstName = 'firstname';
            mockState.lastName = 'lastname';
            const submitButton = await waitFor(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        describe('and the service call is successful', () => {
            it('should redirect to the home page', async () => {

                getUserSpy.mockReturnValueOnce(Promise.resolve(
                    { id: 'id', username: 'un', email: 'email', firstName: 'testSimpleName', lastName: 'testTitle', uuid: 'testUUID', enabled: true }
                ));
                await waitFor(() => {
                    expect(getUserSpy).toHaveBeenCalled();
                    expect(updateUserSpy).toHaveBeenCalled();
                    expect(history.push).toHaveBeenCalledWith(`/user-view/${userId}`, { successMessage: 'The user was amended successfully' });
                });
            });
            it('should call the begin submit action', async () => {

                await waitFor(() => {
                    expect(clearErrorsSpy).toHaveBeenCalled();
                });
            });
        });
    });
    describe('and the data is not filled in', () => {
        beforeEach(async () => {
            updateUserSpy.mockReturnValueOnce(Promise.resolve(
                { id: 'id', username: 'un', email: 'email', firstName: '', lastName: 'testTitle', uuid: 'testUUID', enabled: true }
            ));
            const submitButton = await waitFor(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        it('should call the begin submit action', () => {
            expect(clearErrorsSpy).toHaveBeenCalled();
        });

        it('should set the error state', () => {
            expect(addFormErrorSpy).toHaveBeenNthCalledWith(1, { key: 'firstName', value: 'The First Name is required' });
        });
    });

    describe('and the data is filled in', () => {

        beforeEach(async () => {
            updateUserSpy.mockReset();
            updateUserSpy.mockImplementation(() => Promise.reject({ response: { status: 500 } }));
            mockState.firstName = 'firstname';
            mockState.lastName = 'lastname';
            const submitButton = await waitFor(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        describe('and the service call fails', () => {

            it('should set the error state', () => {
                expect(setMessageSpy).toHaveBeenCalledWith({ description: AMEND_USER_ERROR_DESCRIPTION, title: AMEND_USER_ERROR_TITLE });
            });

            it('should call the begin submit action', () => {
                expect(clearErrorsSpy).toHaveBeenCalled();
            });
        });
    });

    describe('and the data is filled in with invalid data', () => {

        const errorMessage = 'errMsg';

        beforeEach(async () => {
            updateUserSpy.mockReset();
            updateUserSpy.mockImplementation(() => Promise.reject({ response: { status: 400, data: { message: errorMessage } } }));
            mockState.firstName = 'firstname';
            mockState.lastName = 'lastname';
            const submitButton = await waitFor(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        describe('and the service call fails with a bad request', () => {

            it('should set the error state', () => {
                expect(setMessageSpy).toHaveBeenCalledWith({ description: errorMessage, title: AMEND_USER_BAD_REQUEST_TITLE });
            });

            it('should call the begin submit action', () => {
                expect(clearErrorsSpy).toHaveBeenCalled();
            });
        });
    });
});
