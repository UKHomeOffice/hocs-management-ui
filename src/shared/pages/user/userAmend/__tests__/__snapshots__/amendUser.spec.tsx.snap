// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the AmendUser component is mounted should render with default props 1`] = `
<div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-two-thirds-from-desktop"
    >
      <a
        class="govuk-back-link"
        href="/user-view/xx-xx-xx-xx"
      >
        Back
      </a>
      <h1
        class="govuk-heading-xl"
      >
        Amend User Details
      </h1>
    </div>
  </div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-one-half-from-desktop"
    >
      <form
        method="PUT"
      >
        <input
          name="_csrf"
          type="hidden"
          value=""
        />
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="username"
            id="username-label"
          >
            Username
          </label>
          <input
            class="govuk-input"
            disabled=""
            id="username"
            name="username"
            type="text"
            value=""
          />
        </div>
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="email"
            id="email-label"
          >
            Email Address
          </label>
          <input
            class="govuk-input"
            disabled=""
            id="email"
            name="email"
            type="text"
            value=""
          />
        </div>
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="firstName"
            id="firstName-label"
          >
            First Name
          </label>
          <input
            class="govuk-input"
            id="firstName"
            name="firstName"
            type="text"
            value=""
          />
        </div>
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="lastName"
            id="lastName-label"
          >
            Last Name
          </label>
          <input
            class="govuk-input"
            id="lastName"
            name="lastName"
            type="text"
            value=""
          />
        </div>
        <input
          class="govuk-button"
          type="submit"
          value="Submit"
        />
      </form>
    </div>
  </div>
</div>
`;
